### PR TITLE
Don't warn about unsupported RoomType values

### LINKED ISSUE #790: Custom room types shouldn't log spam
So I started using libQuotient in a project and it uses a custom room type for some functionality. However, libQuotient is *not* pleased about this at all. It will remind me constantly in another libQutoient consumer, for example NeoChat:

```
2024-08-28T15:14:30 [WARN ] quotient.events: /home/josh/kde/src/libquotient/Quotient/converters.cpp:13: void Quotient::_impl::warnUnknownEnumValue(const QString&, const char*): Unknown RoomType value: zone.xiv.astra-sync
2024-08-28T15:14:30 [WARN ] quotient.events: /home/josh/kde/src/libquotient/Quotient/converters.cpp:13: void Quotient::_impl::warnUnknownEnumValue(const QString&, const char*): Unknown RoomType value: zone.xiv.astra-sync
2024-08-28T15:14:30 [WARN ] quotient.events: /home/josh/kde/src/libquotient/Quotient/converters.cpp:13: void Quotient::_impl::warnUnknownEnumValue(const QString&, const char*): Unknown RoomType value: zone.xiv.astra-sync
2024-08-28T15:14:30 [WARN ] quotient.events: /home/josh/kde/src/libquotient/Quotient/converters.cpp:13: void Quotient::_impl::warnUnknownEnumValue(const QString&, const char*): Unknown RoomType value: zone.xiv.astra-sync
2024-08-28T15:14:30 [WARN ] quotient.events: /home/josh/kde/src/libquotient/Quotient/converters.cpp:13: void Quotient::_impl::warnUnknownEnumValue(const QString&, const char*): Unknown RoomType value: zone.xiv.astra-sync
2024-08-28T15:14:30 [WARN ] quotient.events: /home/josh/kde/src/libquotient/Quotient/converters.cpp:13: void Quotient::_impl::warnUnknownEnumValue(const QString&, const char*): Unknown RoomType value: zone.xiv.astra-sync
...
(a bunch more)
```

I can easily fix this, but I don't exactly how I should. Would it be enough to change this from a warning to a debug message or?

### PR DESCRIPTION
**Pull request by redstrate**
enumFromJsonString is changed to return a std::optional, and lets the callee handle returning a default value and warning if they want to.

Fixes #790

### CODE REVIEW

File: Quotient/events/roomcreateevent.cpp [hunk 1]
```diff
@@ -8,8 +8,7 @@ using namespace Quotient;
 template <>
 RoomType Quotient::fromJson(const QJsonValue& jv)
 {
-    return enumFromJsonString(jv.toString(), RoomTypeStrings,
-                              RoomType::Undefined);
+    return enumFromJsonString<RoomType>(jv.toString(), RoomTypeStrings).or_else([] -> std::optional<RoomType> { return RoomType::Undefined; }).value();
```
Comment (KitsuneRal) [inline Quotient/events/roomcreateevent.cpp]:
```suggestion
    return enumFromJsonString<RoomType>(jv.toString(), RoomTypeStrings).value_or(RoomType::Undefined);
```

**Review by KitsuneRal — CHANGES_REQUESTED** [review]
Can you apply the same change to `flagFromJsonString()` too, please (and change `JsonConverter<Membership>` accordingly, dropping the warning from it too)?

File: Quotient/events/roomcreateevent.cpp (hunk 1 above)
Comment (redstrate) [reply Quotient/events/roomcreateevent.cpp]: even better, thanks

**Comment by redstrate** [comment]
> Can you apply the same change to `flagFromJsonString()` too, please (and change `JsonConverter<Membership>` accordingly, dropping the warning from it too)?

Done. That's the last usage of `_impl::warnUnknownEnumValue`, should I kill that or keep it?

**Comment by KitsuneRal** [comment]
> > Can you apply the same change to `flagFromJsonString()` too, please (and change `JsonConverter<Membership>` accordingly, dropping the warning from it too)?
> 
> Done. That's the last usage of `_impl::warnUnknownEnumValue`, should I kill that or keep it?

Kill it!

**Comment by redstrate** [comment]
> Kill it!

It's dead now!

**Comment by KitsuneRal** [comment]
> > Kill it!
> 
> It's dead now!

It has ceased to be! It's expired and gone to meet it's maker! Bereft of life, it rests in peace! This is an ex-function!

**Review by KitsuneRal — APPROVED** [review]

**Pull request merged**